diff -uNar u-boot.origin/include/configs/aspeed-common.h u-boot/include/configs/aspeed-common.h
--- u-boot.origin/include/configs/aspeed-common.h	2020-12-14 16:13:23.502807316 -0500
+++ u-boot/include/configs/aspeed-common.h	2020-12-14 16:54:33.679787640 -0500
@@ -11,13 +11,6 @@
 
 #include <asm/arch/platform.h>
 
-#define CONFIG_BOOTFILE		"all.bin"
-
-#define CONFIG_GATEWAYIP	192.168.0.1
-#define CONFIG_NETMASK		255.255.255.0
-#define CONFIG_IPADDR		192.168.0.45
-#define CONFIG_SERVERIP		192.168.0.81
-
 #define CONFIG_STANDALONE_LOAD_ADDR 0x83000000
 
 /* Misc CPU related */
@@ -43,6 +36,14 @@
 #define CONFIG_SYS_BOOTMAPSZ		(256 * 1024 * 1024)
 #define CONFIG_SYS_MALLOC_LEN		(32 << 20)
 
+/* BigPhys Area*/
+#ifdef CONFIG_SPX_FEATURE_GLOBAL_FRAME_BUFFER_SUPPORT
+#define CONFIG_BIGPHYSAREA          "8192"
+#define CONFIG_BOOTLOGO_SUPPORT     1 //Show Logo during Booting
+#define CONFIG_DISPLAY_ADDRESS      0x83000000
+#else
+#define CONFIG_BIGPHYSAREA          "6144"
+#endif
 /*
  * BOOTP options
  */
@@ -52,9 +53,16 @@
  * Miscellaneous configurable options
  */
 #define CONFIG_SYS_BOOTM_LEN 		(0x800000 * 2)
+#define CONFIG_BOOTCOMMAND		"bootfmh"
+#define CONFIG_ENV_OVERWRITE
+
+#define CONFIG_SYS_BOOTMAPSZ       (256 * 1024 * 1024)
+#define CONFIG_SYS_BOOTM_LEN       (0x800000 * 2)
 
 #define CONFIG_EXTRA_ENV_SETTINGS \
+	"autoload=no\0"			\
 	"verify=yes\0"	\
+	"spi_dma=no\0" \
 	""
 
 /* Environment */
