Index: gcc-8-8-20171108-1.1/src/libgo/go/net/main_unix_test.go
===================================================================
--- gcc-8-8-20171108-1.1.orig/src/libgo/go/net/main_unix_test.go
+++ gcc-8-8-20171108-1.1/src/libgo/go/net/main_unix_test.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris
 
 package net
 
Index: gcc-8-8-20171108-1.1/src/libgo/go/os/exec/lp_unix_test.go
===================================================================
--- gcc-8-8-20171108-1.1.orig/src/libgo/go/os/exec/lp_unix_test.go
+++ gcc-8-8-20171108-1.1/src/libgo/go/os/exec/lp_unix_test.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build darwin dragonfly freebsd linux netbsd openbsd solaris
+// +build darwin dragonfly freebsd gnu linux netbsd openbsd solaris
 
 package exec
 
Index: gcc-8-8-20171108-1.1/src/libgo/go/os/os_unix_test.go
===================================================================
--- gcc-8-8-20171108-1.1.orig/src/libgo/go/os/os_unix_test.go
+++ gcc-8-8-20171108-1.1/src/libgo/go/os/os_unix_test.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build darwin dragonfly freebsd linux netbsd openbsd solaris
+// +build darwin dragonfly freebsd gnu linux netbsd openbsd solaris
 
 package os_test
 
Index: gcc-8-8-20171108-1.1/src/libgo/go/os/signal/signal_test.go
===================================================================
--- gcc-8-8-20171108-1.1.orig/src/libgo/go/os/signal/signal_test.go
+++ gcc-8-8-20171108-1.1/src/libgo/go/os/signal/signal_test.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux netbsd openbsd solaris
 
 package signal
 
Index: gcc-8-8-20171108-1.1/src/libgo/go/runtime/crash_unix_test.go
===================================================================
--- gcc-8-8-20171108-1.1.orig/src/libgo/go/runtime/crash_unix_test.go
+++ gcc-8-8-20171108-1.1/src/libgo/go/runtime/crash_unix_test.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux netbsd openbsd solaris
 
 package runtime_test
 
Index: gcc-8-8-20171108-1.1/src/libgo/go/syscall/exec_unix_test.go
===================================================================
--- gcc-8-8-20171108-1.1.orig/src/libgo/go/syscall/exec_unix_test.go
+++ gcc-8-8-20171108-1.1/src/libgo/go/syscall/exec_unix_test.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux netbsd openbsd solaris
 
 package syscall_test
 
Index: gcc-8-8-20171108-1.1/src/libgo/go/runtime/runtime_unix_test.go
===================================================================
--- gcc-8-8-20171108-1.1.orig/src/libgo/go/runtime/runtime_unix_test.go
+++ gcc-8-8-20171108-1.1/src/libgo/go/runtime/runtime_unix_test.go
@@ -6,7 +6,7 @@
 // We need a fast system call to provoke the race,
 // and Close(-1) is nearly universally fast.
 
-// +build aix darwin dragonfly freebsd linux netbsd openbsd plan9
+// +build aix darwin dragonfly freebsd gnu linux netbsd openbsd plan9
 
 package runtime_test
 
Index: gcc-8-8-20171108-1.1/src/libgo/go/syscall/export_unix_test.go
===================================================================
--- gcc-8-8-20171108-1.1.orig/src/libgo/go/syscall/export_unix_test.go
+++ gcc-8-8-20171108-1.1/src/libgo/go/syscall/export_unix_test.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux netbsd openbsd solaris
 
 package syscall
 
Index: gcc-8-8-20171108-1.1/src/libgo/go/syscall/mmap_unix_test.go
===================================================================
--- gcc-8-8-20171108-1.1.orig/src/libgo/go/syscall/mmap_unix_test.go
+++ gcc-8-8-20171108-1.1/src/libgo/go/syscall/mmap_unix_test.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux netbsd openbsd
+// +build aix darwin dragonfly freebsd gnu linux netbsd openbsd
 
 package syscall_test
 
Index: gcc-8-8-20171108-1.1/src/libgo/go/net/addrselect_test.go
===================================================================
--- gcc-8-8-20171108-1.1.orig/src/libgo/go/net/addrselect_test.go
+++ gcc-8-8-20171108-1.1/src/libgo/go/net/addrselect_test.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build darwin dragonfly freebsd linux netbsd openbsd solaris
+// +build darwin dragonfly freebsd gnu linux netbsd openbsd solaris
 
 package net
 
Index: gcc-8-8-20171108-1.1/src/libgo/go/net/cgo_unix_test.go
===================================================================
--- gcc-8-8-20171108-1.1.orig/src/libgo/go/net/cgo_unix_test.go
+++ gcc-8-8-20171108-1.1/src/libgo/go/net/cgo_unix_test.go
@@ -3,7 +3,7 @@
 // license that can be found in the LICENSE file.
 
 // +build cgo,!netgo
-// +build aix darwin dragonfly freebsd linux netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux netbsd openbsd solaris
 
 package net
 
Index: gcc-8-8-20171108-1.1/src/libgo/go/net/conf_test.go
===================================================================
--- gcc-8-8-20171108-1.1.orig/src/libgo/go/net/conf_test.go
+++ gcc-8-8-20171108-1.1/src/libgo/go/net/conf_test.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build darwin dragonfly freebsd linux netbsd openbsd solaris
+// +build darwin dragonfly freebsd gnu linux netbsd openbsd solaris
 
 package net
 
Index: gcc-8-8-20171108-1.1/src/libgo/go/net/dnsconfig_unix_test.go
===================================================================
--- gcc-8-8-20171108-1.1.orig/src/libgo/go/net/dnsconfig_unix_test.go
+++ gcc-8-8-20171108-1.1/src/libgo/go/net/dnsconfig_unix_test.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux netbsd openbsd solaris
 
 package net
 
Index: gcc-8-8-20171108-1.1/src/libgo/go/net/nss_test.go
===================================================================
--- gcc-8-8-20171108-1.1.orig/src/libgo/go/net/nss_test.go
+++ gcc-8-8-20171108-1.1/src/libgo/go/net/nss_test.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build darwin dragonfly freebsd linux netbsd openbsd solaris
+// +build darwin dragonfly freebsd gnu linux netbsd openbsd solaris
 
 package net
 
Index: gcc-8-8-20171108-1.1/src/libgo/go/cmd/go/go_unix_test.go
===================================================================
--- gcc-8-8-20171108-1.1.orig/src/libgo/go/cmd/go/go_unix_test.go
+++ gcc-8-8-20171108-1.1/src/libgo/go/cmd/go/go_unix_test.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build darwin dragonfly freebsd linux netbsd openbsd solaris
+// +build darwin dragonfly freebsd gnu linux netbsd openbsd solaris
 
 package main_test
 
Index: gcc-8-8-20171108-1.1/src/libgo/go/net/dial_unix_test.go
===================================================================
--- gcc-8-8-20171108-1.1.orig/src/libgo/go/net/dial_unix_test.go
+++ gcc-8-8-20171108-1.1/src/libgo/go/net/dial_unix_test.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux netbsd openbsd solaris
 
 package net
 
Index: gcc-8-8-20171108-1.1/src/libgo/go/net/main_cloexec_test.go
===================================================================
--- gcc-8-8-20171108-1.1.orig/src/libgo/go/net/main_cloexec_test.go
+++ gcc-8-8-20171108-1.1/src/libgo/go/net/main_cloexec_test.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build dragonfly freebsd linux
+// +build dragonfly freebsd gnu linux
 
 package net
 
Index: gcc-8-8-20171108-1.1/src/libgo/go/net/dnsclient_unix_test.go
===================================================================
--- gcc-8-8-20171108-1.1.orig/src/libgo/go/net/dnsclient_unix_test.go
+++ gcc-8-8-20171108-1.1/src/libgo/go/net/dnsclient_unix_test.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux netbsd openbsd solaris
 
 package net
 
Index: gcc-8-8-20171108-1.1/src/libgo/go/os/env_unix_test.go
===================================================================
--- gcc-8-8-20171108-1.1.orig/src/libgo/go/os/env_unix_test.go
+++ gcc-8-8-20171108-1.1/src/libgo/go/os/env_unix_test.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build darwin dragonfly freebsd linux netbsd openbsd solaris
+// +build darwin dragonfly freebsd gnu linux netbsd openbsd solaris
 
 package os_test
 
Index: gcc-8-8-20171108-1.1/src/libgo/go/os/error_unix_test.go
===================================================================
--- gcc-8-8-20171108-1.1.orig/src/libgo/go/os/error_unix_test.go
+++ gcc-8-8-20171108-1.1/src/libgo/go/os/error_unix_test.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build darwin dragonfly freebsd linux nacl netbsd openbsd solaris
+// +build darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris
 
 package os_test
 
Index: gcc-8-8-20171108-1.1/src/libgo/go/internal/poll/export_posix_test.go
===================================================================
--- gcc-8-8-20171108-1.1.orig/src/libgo/go/internal/poll/export_posix_test.go
+++ gcc-8-8-20171108-1.1/src/libgo/go/internal/poll/export_posix_test.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris windows
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris windows
 
 // Export guts for testing on posix.
 // Since testing imports os and os imports internal/poll,
Index: gcc-8-8-20171108-1.1/src/libgo/go/runtime/export_unix_test.go
===================================================================
--- gcc-8-8-20171108-1.1.orig/src/libgo/go/runtime/export_unix_test.go
+++ gcc-8-8-20171108-1.1/src/libgo/go/runtime/export_unix_test.go
@@ -1,8 +1,9 @@
+
 // Copyright 2017 The Go Authors. All rights reserved.
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux netbsd openbsd solaris
 
 package runtime
 
Index: gcc-8-8-20171108-1.1/src/libgo/go/internal/poll/fd_posix_test.go
===================================================================
--- gcc-8-8-20171108-1.1.orig/src/libgo/go/internal/poll/fd_posix_test.go
+++ gcc-8-8-20171108-1.1/src/libgo/go/internal/poll/fd_posix_test.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris windows
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris windows
 
 package poll_test
 
