--- a/src/gcc/config/ia64/ia64.c	2018-01-03 11:03:58.000000000 +0100
+++ b/src/gcc/config/ia64/ia64.c	2018-12-16 12:19:05.420184086 +0100
@@ -6122,13 +6122,6 @@
 static void
 ia64_override_options_after_change (void)
 {
-  if (optimize >= 3
-      && !global_options_set.x_flag_selective_scheduling
-      && !global_options_set.x_flag_selective_scheduling2)
-    {
-      flag_selective_scheduling2 = 1;
-      flag_sel_sched_pipelining = 1;
-    }
   if (mflag_sched_control_spec == 2)
     {
       /* Control speculation is on by default for the selective scheduler,
