From: intrigeri <intrigeri@boum.org>
Date: Wed, 25 Jul 2018 08:36:00 +0000
Subject: aa-notify: point to Debian documentation

Forwarded: not-needed
Bug-Debian: https://bugs.debian.org/904436
---
 utils/notify.conf | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/utils/notify.conf b/utils/notify.conf
index 0385483..df3ce6c 100644
--- a/utils/notify.conf
+++ b/utils/notify.conf
@@ -19,4 +19,4 @@ show_notifications="yes"
 # message_body="This is a custom notification message."
 
 # OPTIONAL - custom notification message footer
-# message_footer="For more information visit https://foo.com"
+message_footer="For more information, visit https://wiki.debian.org/AppArmor"
