From: Tomasz Buchert <tomasz@debian.org>
Date: Tue, 2 Jan 2018 11:09:26 +0100
Subject: Make fetch-ocsp-response use python3

---
 script/fetch-ocsp-response | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/script/fetch-ocsp-response b/script/fetch-ocsp-response
index 0ff7461..185116b 100755
--- a/script/fetch-ocsp-response
+++ b/script/fetch-ocsp-response
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python3
 # -*- coding: utf-8 -*-
 
 # nghttp2 - HTTP/2 C Library
