Index: b/src/libgo/go/cmd/go/internal/base/signal_unix.go
===================================================================
--- a/src/libgo/go/cmd/go/internal/base/signal_unix.go
+++ b/src/libgo/go/cmd/go/internal/base/signal_unix.go
@@ -1,8 +1,9 @@
+
 // Copyright 2012 The Go Authors. All rights reserved.
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris
 
 package base
 
Index: b/src/libgo/go/crypto/x509/root_unix.go
===================================================================
--- a/src/libgo/go/crypto/x509/root_unix.go
+++ b/src/libgo/go/crypto/x509/root_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix dragonfly freebsd linux nacl netbsd openbsd solaris
+// +build aix dragonfly freebsd gnu linux nacl netbsd openbsd solaris
 
 package x509
 
Index: b/src/libgo/go/net/interface_stub.go
===================================================================
--- a/src/libgo/go/net/interface_stub.go
+++ b/src/libgo/go/net/interface_stub.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix nacl
+// +build aix nacl gnu
 
 package net
 
Index: b/src/libgo/go/net/internal/socktest/switch_unix.go
===================================================================
--- a/src/libgo/go/net/internal/socktest/switch_unix.go
+++ b/src/libgo/go/net/internal/socktest/switch_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris
 
 package socktest
 
Index: b/src/libgo/go/net/port_unix.go
===================================================================
--- a/src/libgo/go/net/port_unix.go
+++ b/src/libgo/go/net/port_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux netbsd openbsd solaris nacl
+// +build aix darwin dragonfly freebsd gnu linux netbsd openbsd solaris nacl
 
 // Read system port mappings from /etc/services
 
Index: b/src/libgo/go/os/dir_largefile.go
===================================================================
--- a/src/libgo/go/os/dir_largefile.go
+++ b/src/libgo/go/os/dir_largefile.go
@@ -5,7 +5,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix linux solaris,386 solaris,sparc
+// +build aix gnu linux solaris,386 solaris,sparc
 
 package os
 
Index: b/src/libgo/go/os/dir_regfile.go
===================================================================
--- a/src/libgo/go/os/dir_regfile.go
+++ b/src/libgo/go/os/dir_regfile.go
@@ -6,6 +6,7 @@
 // license that can be found in the LICENSE file.
 
 // +build !aix
+// +build !gnu
 // +build !linux
 // +build !solaris !386
 // +build !solaris !sparc
Index: b/src/libgo/go/os/dir_unix.go
===================================================================
--- a/src/libgo/go/os/dir_unix.go
+++ b/src/libgo/go/os/dir_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris
 
 package os
 
Index: b/src/libgo/go/os/exec_unix.go
===================================================================
--- a/src/libgo/go/os/exec_unix.go
+++ b/src/libgo/go/os/exec_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris
 
 package os
 
Index: b/src/libgo/go/os/stat_atim.go
===================================================================
--- a/src/libgo/go/os/stat_atim.go
+++ b/src/libgo/go/os/stat_atim.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build linux openbsd solaristag
+// +build gnu linux openbsd solaristag
 
 package os
 
Index: b/src/libgo/go/os/stat_unix.go
===================================================================
--- a/src/libgo/go/os/stat_unix.go
+++ b/src/libgo/go/os/stat_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris
 
 package os
 
Index: b/src/libgo/go/os/sys_uname.go
===================================================================
--- a/src/libgo/go/os/sys_uname.go
+++ b/src/libgo/go/os/sys_uname.go
@@ -4,7 +4,7 @@
 
 // For systems which only store the hostname in uname (Solaris).
 
-// +build aix solaris irix rtems
+// +build aix gnu solaris irix rtems
 
 package os
 
Index: b/src/libgo/go/os/user/listgroups_unix.go
===================================================================
--- a/src/libgo/go/os/user/listgroups_unix.go
+++ b/src/libgo/go/os/user/listgroups_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build dragonfly darwin freebsd !android,linux netbsd openbsd
+// +build dragonfly darwin freebsd gnu !android,linux netbsd openbsd
 
 package user
 
Index: b/src/libgo/go/os/wait_unimp.go
===================================================================
--- a/src/libgo/go/os/wait_unimp.go
+++ b/src/libgo/go/os/wait_unimp.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly nacl netbsd openbsd solaris
+// +build aix darwin dragonfly gnu nacl netbsd openbsd solaris
 
 package os
 
Index: b/src/libgo/go/path/filepath/path_unix.go
===================================================================
--- a/src/libgo/go/path/filepath/path_unix.go
+++ b/src/libgo/go/path/filepath/path_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris
 
 package filepath
 
Index: b/src/libgo/go/runtime/env_posix.go
===================================================================
--- a/src/libgo/go/runtime/env_posix.go
+++ b/src/libgo/go/runtime/env_posix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris windows
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris windows
 
 package runtime
 
Index: b/src/libgo/go/runtime/lock_sema.go
===================================================================
--- a/src/libgo/go/runtime/lock_sema.go
+++ b/src/libgo/go/runtime/lock_sema.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin nacl netbsd openbsd plan9 solaris windows
+// +build aix darwin gnu nacl netbsd openbsd plan9 solaris windows
 
 package runtime
 
Index: b/src/libgo/go/runtime/netpoll.go
===================================================================
--- a/src/libgo/go/runtime/netpoll.go
+++ b/src/libgo/go/runtime/netpoll.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris windows
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris windows
 
 package runtime
 
Index: b/src/libgo/go/runtime/signal_gccgo.go
===================================================================
--- a/src/libgo/go/runtime/signal_gccgo.go
+++ b/src/libgo/go/runtime/signal_gccgo.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux netbsd openbsd solaris
 
 package runtime
 
Index: b/src/libgo/go/runtime/signal_sighandler.go
===================================================================
--- a/src/libgo/go/runtime/signal_sighandler.go
+++ b/src/libgo/go/runtime/signal_sighandler.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris
 
 package runtime
 
Index: b/src/libgo/go/syscall/errstr.go
===================================================================
--- a/src/libgo/go/syscall/errstr.go
+++ b/src/libgo/go/syscall/errstr.go
@@ -4,6 +4,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
+// +build !gnu
 // +build !linux
 
 package syscall
Index: b/src/libgo/go/syscall/libcall_posix.go
===================================================================
--- a/src/libgo/go/syscall/libcall_posix.go
+++ b/src/libgo/go/syscall/libcall_posix.go
@@ -9,6 +9,8 @@
 // Note that sometimes we use a lowercase //sys name and
 // wrap it in our own nicer implementation.
 
+// +build !gnu
+
 package syscall
 
 import "unsafe"
Index: b/src/libgo/go/syscall/libcall_posix_largefile.go
===================================================================
--- a/src/libgo/go/syscall/libcall_posix_largefile.go
+++ b/src/libgo/go/syscall/libcall_posix_largefile.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix linux solaris,386 solaris,sparc
+// +build aix gnu linux solaris,386 solaris,sparc
 
 // POSIX library calls on systems which use the largefile interface.
 
Index: b/src/libgo/go/syscall/libcall_posix_regfile.go
===================================================================
--- a/src/libgo/go/syscall/libcall_posix_regfile.go
+++ b/src/libgo/go/syscall/libcall_posix_regfile.go
@@ -3,6 +3,7 @@
 // license that can be found in the LICENSE file.
 
 // +build !aix
+// +build !gnu
 // +build !linux
 // +build !solaris !386
 // +build !solaris !sparc
Index: b/src/libgo/go/syscall/libcall_posix_utimesnano.go
===================================================================
--- a/src/libgo/go/syscall/libcall_posix_utimesnano.go
+++ b/src/libgo/go/syscall/libcall_posix_utimesnano.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd openbsd netbsd solaris
+// +build aix darwin dragonfly freebsd gnu openbsd netbsd solaris
 
 // General POSIX version of UtimesNano.
 
Index: b/src/libgo/go/net/cgo_resnew.go
===================================================================
--- a/src/libgo/go/net/cgo_resnew.go
+++ b/src/libgo/go/net/cgo_resnew.go
@@ -3,7 +3,7 @@
 // license that can be found in the LICENSE file.
 
 // +build cgo,!netgo
-// +build aix darwin linux,!android netbsd solaris
+// +build aix darwin gnu linux,!android netbsd solaris
 
 package net
 
Index: b/src/libgo/go/net/cgo_sockold.go
===================================================================
--- a/src/libgo/go/net/cgo_sockold.go
+++ b/src/libgo/go/net/cgo_sockold.go
@@ -3,7 +3,7 @@
 // license that can be found in the LICENSE file.
 
 // +build cgo,!netgo
-// +build aix darwin dragonfly freebsd netbsd openbsd
+// +build aix darwin dragonfly freebsd gnu netbsd openbsd
 
 package net
 
Index: b/src/libgo/go/internal/poll/fd_poll_runtime.go
===================================================================
--- a/src/libgo/go/internal/poll/fd_poll_runtime.go
+++ b/src/libgo/go/internal/poll/fd_poll_runtime.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux netbsd openbsd windows solaris
+// +build aix darwin dragonfly freebsd gnu linux netbsd openbsd windows solaris
 
 package poll
 
Index: b/src/libgo/go/internal/poll/hook_cloexec.go
===================================================================
--- a/src/libgo/go/internal/poll/hook_cloexec.go
+++ b/src/libgo/go/internal/poll/hook_cloexec.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build dragonfly freebsd linux
+// +build dragonfly freebsd gnu linux
 
 package poll
 
Index: b/src/libgo/go/internal/poll/fd_posix.go
===================================================================
--- a/src/libgo/go/internal/poll/fd_posix.go
+++ b/src/libgo/go/internal/poll/fd_posix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris windows
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris windows
 
 package poll
 
Index: b/src/libgo/go/net/sock_cloexec.go
===================================================================
--- a/src/libgo/go/net/sock_cloexec.go
+++ b/src/libgo/go/net/sock_cloexec.go
@@ -5,7 +5,7 @@
 // This file implements sysSocket and accept for platforms that
 // provide a fast path for setting SetNonblock and CloseOnExec.
 
-// +build dragonfly freebsd linux
+// +build dragonfly freebsd gnu linux
 
 package net
 
Index: b/src/libgo/go/syscall/exec_unix.go
===================================================================
--- a/src/libgo/go/syscall/exec_unix.go
+++ b/src/libgo/go/syscall/exec_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux netbsd openbsd solaris
 
 // Fork, exec, wait, etc.
 
Index: b/src/libgo/go/os/exec/lp_unix.go
===================================================================
--- a/src/libgo/go/os/exec/lp_unix.go
+++ b/src/libgo/go/os/exec/lp_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris
 
 package exec
 
Index: b/src/libgo/go/os/signal/signal_unix.go
===================================================================
--- a/src/libgo/go/os/signal/signal_unix.go
+++ b/src/libgo/go/os/signal/signal_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris windows
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris windows
 
 package signal
 
Index: b/src/libgo/go/os/error_unix.go
===================================================================
--- a/src/libgo/go/os/error_unix.go
+++ b/src/libgo/go/os/error_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris
 
 package os
 
Index: b/src/libgo/go/os/file_posix.go
===================================================================
--- a/src/libgo/go/os/file_posix.go
+++ b/src/libgo/go/os/file_posix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris windows
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris windows
 
 package os
 
Index: b/src/libgo/go/os/path_unix.go
===================================================================
--- a/src/libgo/go/os/path_unix.go
+++ b/src/libgo/go/os/path_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris
 
 package os
 
Index: b/src/libgo/go/os/sys_unix.go
===================================================================
--- a/src/libgo/go/os/sys_unix.go
+++ b/src/libgo/go/os/sys_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix dragonfly linux netbsd openbsd solaris
+// +build aix dragonfly gnu linux netbsd openbsd solaris
 
 package os
 
Index: b/src/libgo/go/os/user/decls_unix.go
===================================================================
--- a/src/libgo/go/os/user/decls_unix.go
+++ b/src/libgo/go/os/user/decls_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build darwin dragonfly freebsd linux netbsd openbsd
+// +build darwin dragonfly freebsd gnu linux netbsd openbsd
 // +build cgo
 
 package user
Index: b/src/libgo/go/os/user/lookup_unix.go
===================================================================
--- a/src/libgo/go/os/user/lookup_unix.go
+++ b/src/libgo/go/os/user/lookup_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build darwin dragonfly freebsd !android,linux nacl netbsd openbsd solaris
+// +build darwin dragonfly freebsd gnu !android,linux nacl netbsd openbsd solaris
 // +build !cgo
 
 package user
Index: b/src/libgo/go/syscall/env_unix.go
===================================================================
--- a/src/libgo/go/syscall/env_unix.go
+++ b/src/libgo/go/syscall/env_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris
 
 // Unix environment variables.
 
Index: b/src/libgo/go/syscall/exec_bsd.go
===================================================================
--- a/src/libgo/go/syscall/exec_bsd.go
+++ b/src/libgo/go/syscall/exec_bsd.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu netbsd openbsd solaris
 
 package syscall
 
Index: b/src/libgo/go/syscall/sockcmsg_unix.go
===================================================================
--- a/src/libgo/go/syscall/sockcmsg_unix.go
+++ b/src/libgo/go/syscall/sockcmsg_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux netbsd openbsd solaris
 
 // Socket control messages
 
Index: b/src/libgo/go/syscall/syscall_unix.go
===================================================================
--- a/src/libgo/go/syscall/syscall_unix.go
+++ b/src/libgo/go/syscall/syscall_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux netbsd openbsd solaris
 
 package syscall
 
Index: b/src/libgo/go/time/sys_unix.go
===================================================================
--- a/src/libgo/go/time/sys_unix.go
+++ b/src/libgo/go/time/sys_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris
 
 package time
 
Index: b/src/libgo/go/time/zoneinfo_unix.go
===================================================================
--- a/src/libgo/go/time/zoneinfo_unix.go
+++ b/src/libgo/go/time/zoneinfo_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin,386 darwin,amd64 dragonfly freebsd linux,!android nacl netbsd openbsd solaris
+// +build aix darwin,386 darwin,amd64 dragonfly freebsd gnu linux,!android nacl netbsd openbsd solaris
 
 // Parse "zoneinfo" time zone file.
 // This is a fairly standard file format used on OS X, Linux, BSD, Sun, and others.
Index: b/src/libgo/go/net/addrselect.go
===================================================================
--- a/src/libgo/go/net/addrselect.go
+++ b/src/libgo/go/net/addrselect.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux netbsd openbsd solaris
 
 // Minimal RFC 6724 address selection.
 
Index: b/src/libgo/go/net/conf.go
===================================================================
--- a/src/libgo/go/net/conf.go
+++ b/src/libgo/go/net/conf.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux netbsd openbsd solaris
 
 package net
 
Index: b/src/libgo/go/net/dnsclient_unix.go
===================================================================
--- a/src/libgo/go/net/dnsclient_unix.go
+++ b/src/libgo/go/net/dnsclient_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux netbsd openbsd solaris
 
 // DNS client: see RFC 1035.
 // Has to be linked into package net for Dial.
Index: b/src/libgo/go/net/dnsconfig_unix.go
===================================================================
--- a/src/libgo/go/net/dnsconfig_unix.go
+++ b/src/libgo/go/net/dnsconfig_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux netbsd openbsd solaris
 
 // Read system DNS config from /etc/resolv.conf
 
Index: b/src/libgo/go/net/fd_unix.go
===================================================================
--- a/src/libgo/go/net/fd_unix.go
+++ b/src/libgo/go/net/fd_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris
 
 package net
 
Index: b/src/libgo/go/net/file_unix.go
===================================================================
--- a/src/libgo/go/net/file_unix.go
+++ b/src/libgo/go/net/file_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux netbsd openbsd solaris
 
 package net
 
Index: b/src/libgo/go/net/hook_unix.go
===================================================================
--- a/src/libgo/go/net/hook_unix.go
+++ b/src/libgo/go/net/hook_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris
 
 package net
 
Index: b/src/libgo/go/net/iprawsock_posix.go
===================================================================
--- a/src/libgo/go/net/iprawsock_posix.go
+++ b/src/libgo/go/net/iprawsock_posix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris windows
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris windows
 
 package net
 
Index: b/src/libgo/go/net/ipsock_posix.go
===================================================================
--- a/src/libgo/go/net/ipsock_posix.go
+++ b/src/libgo/go/net/ipsock_posix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris windows
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris windows
 
 package net
 
Index: b/src/libgo/go/net/lookup_unix.go
===================================================================
--- a/src/libgo/go/net/lookup_unix.go
+++ b/src/libgo/go/net/lookup_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux netbsd openbsd solaris
 
 package net
 
Index: b/src/libgo/go/net/nss.go
===================================================================
--- a/src/libgo/go/net/nss.go
+++ b/src/libgo/go/net/nss.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux netbsd openbsd solaris
 
 package net
 
Index: b/src/libgo/go/net/sockopt_posix.go
===================================================================
--- a/src/libgo/go/net/sockopt_posix.go
+++ b/src/libgo/go/net/sockopt_posix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux netbsd openbsd solaris windows
+// +build aix darwin dragonfly freebsd gnu linux netbsd openbsd solaris windows
 
 package net
 
Index: b/src/libgo/go/net/sock_posix.go
===================================================================
--- a/src/libgo/go/net/sock_posix.go
+++ b/src/libgo/go/net/sock_posix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris windows
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris windows
 
 package net
 
Index: b/src/libgo/go/net/tcpsockopt_posix.go
===================================================================
--- a/src/libgo/go/net/tcpsockopt_posix.go
+++ b/src/libgo/go/net/tcpsockopt_posix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux netbsd openbsd solaris windows
+// +build aix darwin dragonfly freebsd gnu linux netbsd openbsd solaris windows
 
 package net
 
Index: b/src/libgo/go/net/tcpsockopt_unix.go
===================================================================
--- a/src/libgo/go/net/tcpsockopt_unix.go
+++ b/src/libgo/go/net/tcpsockopt_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix freebsd linux netbsd
+// +build aix freebsd gnu linux netbsd
 
 package net
 
Index: b/src/libgo/go/net/tcpsock_posix.go
===================================================================
--- a/src/libgo/go/net/tcpsock_posix.go
+++ b/src/libgo/go/net/tcpsock_posix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris windows
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris windows
 
 package net
 
Index: b/src/libgo/go/net/udpsock_posix.go
===================================================================
--- a/src/libgo/go/net/udpsock_posix.go
+++ b/src/libgo/go/net/udpsock_posix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris windows
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris windows
 
 package net
 
Index: b/src/libgo/go/net/unixsock_posix.go
===================================================================
--- a/src/libgo/go/net/unixsock_posix.go
+++ b/src/libgo/go/net/unixsock_posix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris windows
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris windows
 
 package net
 
Index: b/src/libgo/go/os/exec_posix.go
===================================================================
--- a/src/libgo/go/os/exec_posix.go
+++ b/src/libgo/go/os/exec_posix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris windows
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris windows
 
 package os
 
Index: b/src/libgo/go/os/file_unix.go
===================================================================
--- a/src/libgo/go/os/file_unix.go
+++ b/src/libgo/go/os/file_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris
 
 package os
 
Index: b/src/libgo/go/os/executable_procfs.go
===================================================================
--- a/src/libgo/go/os/executable_procfs.go
+++ b/src/libgo/go/os/executable_procfs.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build linux netbsd dragonfly nacl
+// +build gnu linux netbsd dragonfly nacl
 
 package os
 
Index: b/src/libgo/go/syscall/timestruct.go
===================================================================
--- a/src/libgo/go/syscall/timestruct.go
+++ b/src/libgo/go/syscall/timestruct.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris
 
 package syscall
 
Index: b/src/libgo/go/net/cgo_unix.go
===================================================================
--- a/src/libgo/go/net/cgo_unix.go
+++ b/src/libgo/go/net/cgo_unix.go
@@ -3,7 +3,7 @@
 // license that can be found in the LICENSE file.
 
 // +build cgo,!netgo
-// +build aix darwin dragonfly freebsd linux netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux netbsd openbsd solaris
 
 package net
 
Index: b/src/libgo/go/archive/tar/stat_unix.go
===================================================================
--- a/src/libgo/go/archive/tar/stat_unix.go
+++ b/src/libgo/go/archive/tar/stat_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix linux darwin dragonfly freebsd openbsd netbsd solaris
+// +build aix gnu linux darwin dragonfly freebsd openbsd netbsd solaris
 
 package tar
 
Index: b/src/libgo/go/crypto/rand/eagain.go
===================================================================
--- a/src/libgo/go/crypto/rand/eagain.go
+++ b/src/libgo/go/crypto/rand/eagain.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris
 
 package rand
 
Index: b/src/libgo/go/crypto/rand/rand_unix.go
===================================================================
--- a/src/libgo/go/crypto/rand/rand_unix.go
+++ b/src/libgo/go/crypto/rand/rand_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd plan9 solaris
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd plan9 solaris
 
 // Unix cryptographically secure pseudorandom number
 // generator.
Index: b/src/libgo/go/mime/type_unix.go
===================================================================
--- a/src/libgo/go/mime/type_unix.go
+++ b/src/libgo/go/mime/type_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris
 
 package mime
 
Index: b/src/libgo/go/plugin/plugin_dlopen.go
===================================================================
--- a/src/libgo/go/plugin/plugin_dlopen.go
+++ b/src/libgo/go/plugin/plugin_dlopen.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build linux,cgo darwin,cgo
+// +build linux,cgo darwin,cgo gnu,cgo
 
 package plugin
 
Index: b/src/libgo/go/syscall/dirent.go
===================================================================
--- a/src/libgo/go/syscall/dirent.go
+++ b/src/libgo/go/syscall/dirent.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris
 
 package syscall
 
Index: b/src/libgo/runtime/env_posix.c
===================================================================
--- a/src/libgo/runtime/env_posix.c
+++ b/src/libgo/runtime/env_posix.c
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build darwin dragonfly freebsd linux nacl netbsd openbsd solaris windows
+// +build darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris windows
 
 #include "runtime.h"
 #include "array.h"
Index: b/src/libgo/go/net/internal/socktest/sys_unix.go
===================================================================
--- a/src/libgo/go/net/internal/socktest/sys_unix.go
+++ b/src/libgo/go/net/internal/socktest/sys_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris
 
 package socktest
 
Index: b/src/libgo/go/plugin/plugin_stubs.go
===================================================================
--- a/src/libgo/go/plugin/plugin_stubs.go
+++ b/src/libgo/go/plugin/plugin_stubs.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build !linux,!darwin !cgo
+// +build !gnu !linux,!darwin !cgo
 
 package plugin
 
Index: b/src/libgo/go/net/internal/socktest/sys_cloexec.go
===================================================================
--- a/src/libgo/go/net/internal/socktest/sys_cloexec.go
+++ b/src/libgo/go/net/internal/socktest/sys_cloexec.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build dragonfly freebsd linux
+// +build dragonfly freebsd gnu linux
 
 package socktest
 
Index: b/src/libgo/go/os/user/cgo_lookup_unix.go
===================================================================
--- a/src/libgo/go/os/user/cgo_lookup_unix.go
+++ b/src/libgo/go/os/user/cgo_lookup_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd !android,linux netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu !android,linux netbsd openbsd solaris
 // +build cgo
 
 package user
Index: b/src/libgo/go/net/error_posix.go
===================================================================
--- a/src/libgo/go/net/error_posix.go
+++ b/src/libgo/go/net/error_posix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris windows
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris windows
 
 package net
 
Index: b/src/libgo/go/internal/poll/fd_unix.go
===================================================================
--- a/src/libgo/go/internal/poll/fd_unix.go
+++ b/src/libgo/go/internal/poll/fd_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris
 
 package poll
 
Index: b/src/libgo/go/syscall/forkpipe_bsd.go
===================================================================
--- a/src/libgo/go/syscall/forkpipe_bsd.go
+++ b/src/libgo/go/syscall/forkpipe_bsd.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly netbsd openbsd solaris
+// +build aix darwin dragonfly gnu netbsd openbsd solaris
 
 package syscall
 
Index: b/src/libgo/go/internal/poll/hook_unix.go
===================================================================
--- a/src/libgo/go/internal/poll/hook_unix.go
+++ b/src/libgo/go/internal/poll/hook_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux nacl netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux nacl netbsd openbsd solaris
 
 package poll
 
Index: b/src/libgo/go/internal/poll/sock_cloexec.go
===================================================================
--- a/src/libgo/go/internal/poll/sock_cloexec.go
+++ b/src/libgo/go/internal/poll/sock_cloexec.go
@@ -5,7 +5,7 @@
 // This file implements sysSocket and accept for platforms that
 // provide a fast path for setting SetNonblock and CloseOnExec.
 
-// +build dragonfly freebsd linux
+// +build dragonfly freebsd gnu linux
 
 package poll
 
Index: b/src/libgo/go/internal/poll/sockopt.go
===================================================================
--- a/src/libgo/go/internal/poll/sockopt.go
+++ b/src/libgo/go/internal/poll/sockopt.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux netbsd openbsd solaris windows
+// +build aix darwin dragonfly freebsd gnu linux netbsd openbsd solaris windows
 
 package poll
 
Index: b/src/libgo/go/internal/poll/sockoptip.go
===================================================================
--- a/src/libgo/go/internal/poll/sockoptip.go
+++ b/src/libgo/go/internal/poll/sockoptip.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux netbsd openbsd solaris windows
+// +build aix darwin dragonfly freebsd gnu linux netbsd openbsd solaris windows
 
 package poll
 
Index: b/src/libgo/go/internal/poll/sockopt_unix.go
===================================================================
--- a/src/libgo/go/internal/poll/sockopt_unix.go
+++ b/src/libgo/go/internal/poll/sockopt_unix.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build darwin dragonfly freebsd linux netbsd openbsd solaris
+// +build darwin dragonfly freebsd gnu linux netbsd openbsd solaris
 
 package poll
 
Index: b/src/libgo/go/archive/tar/stat_actime1.go
===================================================================
--- a/src/libgo/go/archive/tar/stat_actime1.go
+++ b/src/libgo/go/archive/tar/stat_actime1.go
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix linux dragonfly openbsd solaris
+// +build aix gnu linux dragonfly openbsd solaris
 
 package tar
 
--- a/src/libgo/go/golang_org/x/net/internal/nettest/helper_nobsd.go	2018-01-10 11:13:47.000000000 +0100
+++ b/src/libgo/go/golang_org/x/net/internal/nettest/helper_nobsd.go	2018-03-02 13:37:11.000000000 +0100
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix linux solaris
+// +build aix gnu linux solaris
 
 package nettest
 
--- a/src/libgo/go/golang_org/x/net/internal/nettest/helper_posix.go	2018-01-10 11:13:47.000000000 +0100
+++ b/src/libgo/go/golang_org/x/net/internal/nettest/helper_posix.go	2018-03-01 23:01:22.000000000 +0100
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux netbsd openbsd solaris windows
+// +build aix darwin dragonfly freebsd gnu linux netbsd openbsd solaris windows
 
 package nettest
 
--- a/src/libgo/go/golang_org/x/net/internal/nettest/helper_unix.go	2018-01-10 11:13:47.000000000 +0100
+++ b/src/libgo/go/golang_org/x/net/internal/nettest/helper_unix.go	2018-03-02 13:40:42.000000000 +0100
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux netbsd openbsd solaris
 
 package nettest
 
--- a/src/libgo/go/net/sockoptip_posix.go	2018-01-09 04:04:37.000000000 +0100
+++ b/src/libgo/go/net/sockoptip_posix.go	2018-03-01 22:55:56.000000000 +0100
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux netbsd openbsd solaris windows
+// +build aix darwin dragonfly freebsd gnu linux netbsd openbsd solaris windows
 
 package net
 
--- a/src/libgo/go/os/signal/internal/pty/pty.go	2018-02-08 18:22:43.000000000 +0100
+++ b/src/libgo/go/os/signal/internal/pty/pty.go	2018-03-01 23:09:21.000000000 +0100
@@ -2,7 +2,7 @@
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-// +build aix darwin dragonfly freebsd linux,!android netbsd openbsd solaris
+// +build aix darwin dragonfly freebsd gnu linux,!android netbsd openbsd solaris
 
 // Package pty is a simple pseudo-terminal package for Unix systems,
 // implemented by calling C functions via cgo.
