# DP: Link with the shared libphobos library by default.

Index: b/src/gcc/d/d-spec.c
===================================================================
--- a/src/gcc/d/d-spec.c
+++ b/src/gcc/d/d-spec.c
@@ -395,9 +395,9 @@ lang_specific_driver (cl_decoded_option
   /* Add `-lgphobos' if we haven't already done so.  */
   if (library > 0 && need_phobos)
     {
-      /* Default to static linking.  */
+      /* Default to shared linking.  */
       if (library == 1)
-        library = 2;
+        library = 3;
 
 #ifdef HAVE_LD_STATIC_DYNAMIC
       if (library == 3 && static_link)
