Index: b/src/libphobos/configure.ac
===================================================================
--- a/src/libphobos/configure.ac
+++ b/src/libphobos/configure.ac
@@ -154,6 +154,8 @@ PHOBOS_SOVERSION="76:2:0"
 AC_SUBST([DRUNTIME_SOVERSION])
 AC_SUBST([PHOBOS_SOVERSION])
 
+# trigger rebuild of the configure file
+
 # Set default flags (after DRUNTIME_WERROR!)
 if test -z "$GDCFLAGS"; then
     GDCFLAGS="-Wall $WERROR_FLAG -g -frelease -O2"
Index: b/src/libphobos/m4/druntime.m4
===================================================================
--- a/src/libphobos/m4/druntime.m4
+++ b/src/libphobos/m4/druntime.m4
@@ -98,6 +98,7 @@ AC_DEFUN([DRUNTIME_INSTALL_DIRECTORIES],
 
   # Default case for install directory for D sources files.
   gdc_include_dir='$(libdir)/gcc/${target_alias}/${gcc_version}/include/d'
+  gdc_include_dir='${libdir}/gcc-cross/${target_alias}'/${gcc_version}/include/d
   AC_SUBST(gdc_include_dir)
 ])
 
