From: Net-SNMP Packaging Team <pkg-net-snmp-devel@lists.alioth.debian.org>
Date: Thu, 18 Jun 2015 06:12:05 +0900
Subject: fix_logging_option

---
 man/snmpd.8.def | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/man/snmpd.8.def b/man/snmpd.8.def
index f173396..d86023a 100644
--- a/man/snmpd.8.def
+++ b/man/snmpd.8.def
@@ -78,9 +78,9 @@ To get a list of compiled modules, run the agent with the arguments
 .I "\-Dmib_init \-H"
 (assuming debugging support has been compiled in).
 .TP
-.B \-L[efos]
+.B \-L[eEfFoOsS]
 Specify where logging output should be directed (standard error or output,
-to a file or via syslog).  See LOGGING OPTIONS in snmpcmd(5) for details.
+to a file or via syslog).  See LOGGING OPTIONS in snmpcmd(1) for details.
 .TP
 .BR \-m " \fIMIBLIST"
 Specifies a colon separated list of MIB modules to load for this
